ibtk/include/ibtk/libmesh_utilities.h

```diff
@@ -350,14 +345,9 @@ get_values_for_interpolation(MultiArray& U_node,
                              const std::vector<std::vector<unsigned int> >& dof_indices)
 {
     libMesh::PetscVector<double>* U_petsc_vec = dynamic_cast<libMesh::PetscVector<double>*>(&U_vec);
-    Vec U_global_vec = U_petsc_vec->vec();
-    Vec U_local_vec;
-    VecGhostGetLocalForm(U_global_vec, &U_local_vec);
-    double* U_local_soln;
-    VecGetArray(U_local_vec, &U_local_soln);
+    const double* const U_local_soln = U_petsc_vec->get_array_read();
     get_values_for_interpolation(U_node, *U_petsc_vec, U_local_soln, dof_indices);
-    VecRestoreArray(U_local_vec, &U_local_soln);
-    VecGhostRestoreLocalForm(U_global_vec, &U_local_vec);
+    U_petsc_vec->restore_array();
     return;
 } // get_values_for_interpolation
```

Comment:
This was the first approach I considered, but it is a lot less efficient than doing the read operation at the beginning of, e.g., `interpWeighted` and restoring at the end since libMesh will destroy the locally ghosted `Vec` when calling `restore_array` by calling `VecRestoreArrayRead` and `VecGhostRestoreLocalForm`.

Doing it this way is somewhat faster than what we have on master, but I think the right approach is to put this kind of localization outside the loop that calls `get_values_for_interpolation` repeatedly to avoid this. The downside is that we have to modify all such loops to get the full performance improvement.

Reply:
Replying by email did not preserve the thread here. To recap:

That is really a convenience function that is not used in any hot code paths (I think!). Because it is slow, I was thinking about deprecating it. I think it is very unlikely to show up in applications, however, and so it also is tempting to just delete it.

I am pretty sure that this particular version of `get_values_for_interpolation` is only used in a couple of places in the code, in code related to computing constraint forces between overlapping structures. Those constraint forces do not work properly, however, and hopefully no one is using them...

Reply:
Thats a good point; we could avoid the problem by just inlining this whole function. I'll take a look.

Reply:
I am just repeating myself here, but note that the version of the function that this comment is referring to _is not_ used in IBAMR except for a couple of places. Most calls to `get_values_for_interpolation` use the version that includes the extracted C-style array (see [here](https://github.com/IBAMR/IBAMR/blob/80cc1551e82308b88fbf64b980b80b915091aef2/ibtk/include/ibtk/libmesh_utilities.h#L286)). I think that all of the calls to that function use that version of the function, and the calls use C arrays that are extracted outside of the element or patch loop.